zerocopy-derive/src/enums.rs

```diff
@@ -0,0 +1,491 @@
+            discriminant: Ident::new(&format!("Raw{}Discriminant", ident), ident.span()),
+            fields: Ident::new(&format!("Raw{}Fields", ident), ident.span()),
+            variant_prefix: format!("Raw{}Variant", ident),
+        }
+    }
+
+    fn discriminant_const(&self, variant: &Ident) -> Ident {
+        Ident::new(&format!("{}{}", Self::DISCRIMINANT_VALUE_PREFIX, variant), variant.span())
+    }
+
+    fn variant_struct(&self, variant: &Ident) -> Ident {
+        Ident::new(&format!("{}{}", self.variant_prefix, variant), variant.span())
+    }
+}
+
+/// Generates an implementation of `is_bit_valid` for the given enum input.
+pub(crate) fn derive_is_bit_valid(
+    reprs: &[EnumRepr],
+    ident: &Ident,
+    generics: &Generics,
+    data: &DataEnum,
+) -> TokenStream {
+    let idents = RawIdents::for_enum(ident);
+
+    if reprs.contains(&EnumRepr::C) {
+        derive_is_bit_valid_c(data, generics, &idents)
+    } else {
+        derive_is_bit_valid_primitive(data, generics, discriminant_repr(reprs), &idents)
+    }
+}
+
+pub(crate) fn discriminant_repr(reprs: &[EnumRepr]) -> &EnumRepr {
+    if reprs.contains(&EnumRepr::C) {
+        reprs.iter().find(|&r| r != &EnumRepr::C).unwrap_or(&EnumRepr::C)
+    } else {
+        reprs.first().unwrap()
+    }
+}
+
+pub(crate) fn generate_discriminant_type(
+    data: &DataEnum,
+    repr: &EnumRepr,
+    name: &Ident,
+) -> TokenStream {
+    let variants = data.variants.iter().map(|v| {
+        let ident = &v.ident;
+        if let Some((eq, discriminant)) = &v.discriminant {
+            quote! { #ident #eq #discriminant }
+        } else {
+            quote! { #ident }
+        }
+    });
+
+    quote! {
+        #[repr(#repr)]
+        #[allow(dead_code)]
+        enum #name {
+            #(#variants),*
+        }
+    }
+}
+
+fn generate_discriminant_consts(data: &DataEnum, idents: &RawIdents) -> TokenStream {
+    let discriminant = &idents.discriminant;
+
+    let consts = data.variants.iter().map(|v| {
+        let ident = &v.ident;
+        let const_ident = idents.discriminant_const(ident);
+        quote! {
+            // SAFETY: The discriminant type is a C-style enum, and all C-style
+            // enums are represented as simple integers. All integers can be
+            // transmuted to their constituent bytes.
+            const #const_ident: [u8; ::core::mem::size_of::<#discriminant>()] = unsafe {
+                ::core::mem::transmute(#discriminant::#ident)
+            };
+        }
+    });
+
+    quote! {
+        #(#consts)*
+    }
+}
+
+fn generate_variant_struct(
+    variant: &Variant,
+    generics: &Generics,
+    idents: &RawIdents,
+    with_tag: bool,
+) -> TokenStream {
+    let discriminant_ident = &idents.discriminant;
+    let variant_struct_ident = idents.variant_struct(&variant.ident);
+
+    let type_params = generics.type_params().map(|p| &p.ident);
+    let phantom_ty = quote! {
+        ::core::marker::PhantomData<(#(#type_params,)*)>
```

Comment:
Here and elsewhere: Can you use `core` via `::zerocopy::macro_util::core_reexport` for hygeine?

Reply:
Leading double colon tells the compiler to look in the external prelude (links [1](https://doc.rust-lang.org/reference/paths.html), [2](https://doc.rust-lang.org/reference/names/preludes.html#extern-prelude)). `core` is always in the external prelude

Reply:
I believe we favor `::zerocopy::macro_util::core_reexport` so that our macros would work inside 2015 edition crates.

Reply:
Done